docs/hacking.rst

```diff
@@ -2,273 +2,256 @@
+additional parameter, ``errors``, which if provided must be an Array to which 
+errors will be pushed.
 
 Example::
 
-  const brandStrings = `brandName = Loki`;
-
-  const strings = `
-
-  hello-world = Hello { $userName }! Welcome to { brandName } 
-  open-pref = { OS() ->
-    [mac] Open Preferences
-   *[other] Open Settings 
-  }
+  // Remove indentation from FTL samples below.
+  const ftl = ([text]) => text.replace(/^\s*/mg, '');
 
+  // Define two translation resources.
+  const brandTranslations = ftl`
+      brand-name = Loki
+  `;
+  const mainTranslations = ftl`
+      hello-world = Hello { $username }! Welcome to { brand-name } 
+      open-pref = { OS() ->
+          [darwin] Open Preferences
+         *[other]  Open Settings 
+      }
   `;
 
+  // Create the MessageContext instance with a custom OS function.
   const mc = new MessageContext('en-US', {
     functions: {
       'OS': function() {
-        return env.platform;
+        return process.platform;
       }
     }
   });
 
-  mc.addMessages(brandStrings);
-  mc.addMessages(strings);
+  // Add and parse the FTL translations.
+  mc.addMessages(brandTranslations);
+  mc.addMessages(mainTranslations);
+
+  // An array for collection formatting errors.
+  const errors = [];
 
   const str1 = mc.format('hello-world', {
-    userName: 'John'
-  });
+    username: 'John'
+  }, errors);
 
-  // str1 === "Hello John! welcome to Loki"
+  // str1 === "Hello John! Welcome to Loki"
+  // errors.length == 0
 
-Localization API
-===================
+  const str2 = mc.format('hello-world', null, errors);
 
-Second layer has two fundamental features over the first one.
-It interacts with environment by using I/O to load resources and it is multi-lignual
-with language fallbacking.
+  // str1 === "Hello username! Welcome to Loki"
+  // errors.length == 1
+  // errors[0] is a <ReferenceError: Unknown external: username>
 
-Due to those two differences, it's fully asynchronous.
 
-ResourceBundle
---------------
+The Localization API
+====================
 
-A bundle is a single-locale object that holds collection of resource sources
-necessary for a given ``Localization`` object and corresponding ``MessageContext`` object.
+The ``Localization`` class represents the main functionality of l20n.js.  It 
+can retrieve formatted translations in accordance to the user's language 
+preferences, resorting to fallback languages in case of errors.  It also 
+interacts with the environment via I/O in order to load resources.  It is the 
+work-horse of l20n.js.  
 
-``ResourceBundle`` API:
+In Web-like runtimes the ``Localization`` API is fully asynchronous.  This 
+allows for loading fallback resources lazily in case of errors in the first 
+language.
 
-:lang:
-    Language tag associated with the bundle
+The ``Localization`` class can be thought of as a multi-lingual, asynchronous 
+wrapper on top of multiple ``MessageContext`` instances which handles lazy-IO 
+and the fallback strategy.
```

Comment:
`language fallback` - I believe that it's worth pointing out what is fallbacking